### PR TITLE
Add IPhysicalProjectTreeStorage.AddFileAsync

### PR DESCRIPTION
**Pull request by davkean**
This is breakout from a larger PR where I'm adding licenses special file provider support, this adds a new method called `IPhysicalProjectTreeStorage.AddFileAsync` which creates a zero'd file on disk and publishes it to Solution Explorer before it has returned.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs [hunk 1]
```diff
@@ -22,34 +23,52 @@ internal class PhysicalProjectTreeStorage : IPhysicalProjectTreeStorage
...
+        public Task<IProjectTree?> CreateFolderAsync(string path)
         {
-            _fileSystem.Value.CreateDirectory(fullPath);
+            Requires.NotNullOrEmpty(path, nameof(path));
+
+            return AddToProjectAsync(path, waitForFileSystemUpdates: false, async fullPath =>
+            {
+                _fileSystem.Value.CreateDirectory(fullPath);
+
+                await _folderManager.Value.IncludeFolderInProjectAsync(fullPath, recursive: false);
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs]: Old code, but should be able to remove this async state machine too

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs [hunk 2]
```diff
@@ -22,34 +23,52 @@ internal class PhysicalProjectTreeStorage : IPhysicalProjectTreeStorage
...
             Requires.NotNullOrEmpty(path, nameof(path));
 
-            if (_treeService.Value.CurrentTree == null)
-                throw new InvalidOperationException("Physical project tree has not yet been published.");
+            return await AddToProjectAsync(path, waitForFileSystemUpdates: true, async fullPath =>
+            {
+                using (_fileSystem.Value.Create(fullPath)) { }
 
-            string fullPath = _unconfiguredProject.MakeRooted(path);
+                await _sourceItemProvider.Value.AddAsync(fullPath);
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs]: Can remote this async state machine, and maybe the outer one, depending on that cast.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IPhysicalProjectTreeStorage.cs [hunk 3]
```diff
@@ -14,11 +14,54 @@ namespace Microsoft.VisualStudio.ProjectSystem
...
     /// <remarks>
     ///     This interface provides a simple facade over the <see cref="IFileSystem"/>, <see cref="IFolderManager"/>, 
-    ///     <see cref="IProjectTreeService"/> and <see cref="IProjectItemProvider"/> interfaces.
+    ///     <see cref="IProjectTreeService"/>, <see cref="IProjectItemProvider"/> and <see cref="IProjectItemProvider"/> interfaces.
     /// </remarks>
     [ProjectSystemContract(ProjectSystemContractScope.UnconfiguredProject, ProjectSystemContractProvider.Private, Cardinality = ImportCardinality.ExactlyOne)]
     internal interface IPhysicalProjectTreeStorage
     {
+        /// <summary>
+        ///     Creates a zeroed file on disk, adding it add to the physical project tree.
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IPhysicalProjectTreeStorage.cs]: Zeroed or empty? To me a zeroed file has non-zero length and contains all zeroes.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs [hunk 4]
```diff
@@ -13,45 +13,58 @@ internal class PhysicalProjectTreeStorage : IPhysicalProjectTreeStorage
...
+
+            }) as IProjectItem;
+        }
+
+        public Task<IProjectTree?> CreateFolderAsync(string path)
+        {
+            Requires.NotNullOrEmpty(path, nameof(path));
 
-            return AddToProjectAsync(fullPath);
+            return AddToProjectAsync(path, waitForFileSystemUpdates: false, async fullPath =>
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs]: `CreateEmpyFileAsync` awaits the `AddToProjectAsync` task, whereas this returns it directly. Perhaps they should be consistent. Awaiting in both will create clearer stack traces.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/SpecialFileProviders/AbstractFindByNameSpecialFileProvider.cs [hunk 5]
```diff
@@ -29,8 +28,8 @@ internal abstract class AbstractFindByNameSpecialFileProvider : ISpecialFileProv
 
         protected AbstractFindByNameSpecialFileProvider(
             IPhysicalProjectTree projectTree,
-            [Import(ExportContractNames.ProjectItemProviders.SourceFiles)] IProjectItemProvider sourceItemsProvider,
-            [Import(AllowDefault = true)] Lazy<ICreateFileFromTemplateService>? templateFileCreationService,
+            IProjectItemProvider sourceItemsProvider,
+            Lazy<ICreateFileFromTemplateService>? templateFileCreationService,
```
Comment (drewnoakes) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/SpecialFileProviders/AbstractFindByNameSpecialFileProvider.cs]: This was only nullable due to `AllowDefault = true`:

```suggestion
            Lazy<ICreateFileFromTemplateService> templateFileCreationService,
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs (hunk 4 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs]: Not convinced by this - one needs the result, one doesn't, and would just add unneeded state machinary.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IPhysicalProjectTreeStorage.cs (hunk 3 above)
Comment (davkean) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/IPhysicalProjectTreeStorage.cs]: Yep, good call.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs (hunk 4 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs]: Ah, I missed the `as` cast at the end.

**Pull request merged**